### DISCUSSION TITLE
esp32cam - updated driver - ready for help & testing.

### DISCUSSION THREAD

**Discussion by btsimonh** [arendst/Tasmota#18732] [Ideas]
Hi All,

I've spent the last few weeks working on the webcam driver.
It now has all the features I could think of easily - the ones that I thought could be used from Berry scripting.
It's very stable on my (notoriously unreliable) cam.

It's here:
https://github.com/btsimonh/Tasmota/blob/webcam2023/tasmota/tasmota_xdrv_driver/xdrv_99_esp32_webcamberry.ino

But I don't want to PR until we've got some feedback, and maybe additional ideas taken into account.

VERY brief overview:
It runs all frame grabbing in a thread, so no more possible conflicts, and also a bit more performance.
I've added TAS commands for some of the functions that were previpously only available to scripts (not to berry...).
I've added some image conversion as TAS commands.
It SHOULD still work with (non-berry) scripting (apart from maybe the USE_WEBCAM #define).

Exxample: you can get actual pixels in berry by
WcGetFrame1  (reads the next frame to buffer)
WcConvertFrame1 5 0 (decode the JPEG buffer to RGB888 at scale 1:1)
Wcgetpicstore1  (returns the actual buffer address, length, width, height, format).
(you can download the raw pixels at this point from http://wc.jpeg?p=1)

You could then draw on the image!....  and then
WcConvertFrame1 4 0 (encode to JPEG)
access the resulting jpeg at http://wc.jpeg?p=1

Or...  you could use the image in Tensorflow....

It also implements three forms of motion detection:
1/ JPEG size change.  No performance Impact (i.e. will run at the cam framerate)
2/ accumulated pixel difference (as per original motion, only accessible through scripts).
3/ pixel difference threshold -> count of pixels over threshold.
The motion detection is somehwat faster (300ms vs 1.3s for a large frame) due to a modified JPEG decode to decode direct to monochrome).

You can access the motion buffers (re-encoded to jpeg) via the browser, and enable a diff buffer.
You can access the motion pixels with wcGetMotionPixels (gives addr and len - for berry?)
If informs berry of motion (webcan.motion is called if present).

If it gets an error at start, the error is displayed in the main menu.
You can turn off and on menu video (streaming takes some bandwidth... if you don't want to, you don't need to).

I still get some 0x105 at soft restart occasionally - the cross to bear from having an AIThinker cam module.  But it's rare.

I did try to get raw pixels from the camera (more efficient if that's all you want), but failed.

So, what would you do with such features in a Berry script?  What's missing?

Any testing, ideas, thoughts welcome.

br,
Simon

**Comment by btsimonh** [comment]
I've added a masking image to motion detection - with 'automask' (sets mask from next N motion sense runs).
The mask pixels are available from a command, so in theory, you could set them from Berry.
However, most likely we'd like to load them from file.  This is something Berry could be good at (determining what to load, and where to put it).  But we'd need some more image manipulation functions.
e.g. currently we can get jpeg from the camera, and berry can read and access the raw data.
Berry can convert such images (in place) to raw pixels, and back to jpeg.
But we can't (at the moment) read a jpeg from disk in berry, and set it into one of the 4 'user' buffers in the webcam driver, ready for manipulation.  And the berry access to image data is a little 'knife and fork) (passing raw addresses and length as command repsonses...).

### enable/set mask
WCsetMotiondetect8 0-nnnnn
WCsetMotiondetect8 0 - free
WCsetMotiondetect8 1 - alloc/create
WCsetMotiondetect8 n - alloc/create and set automask to n, pixel threshold to 10, pixel expansion to 4
WCsetMotiondetect8 n m o - alloc/create and set automask to n, pixel threshold to m, pixel expansion to o

**Comment by btsimonh** [comment]
## Brief comparison of the two drivers:
### tasmota32-webcam 13.1.0 - current release
Params used: AIThinker style, wcresolution 5, wcclock 20
Web: ~20fps
Single web session only.
Single RTSP stream only.

Motion only acessible by using scripts (special build).
Image save only accessible using scripts.
Only one form of motion detection available.

### tasmota32-webcam [PR](https://github.com/arendst/Tasmota/pull/19280)
PR binary: [tasmota32-webcam-PR.zip](https://github.com/arendst/Tasmota/files/12387695/tasmota32-webcam-PR.zip)
[Docs](https://github.com/btsimonh/TasmotaBerryPLay/blob/master/tasdocs/esp32cam.md)
[Example Timelapse and Motion](https://github.com/btsimonh/TasmotaBerryPLay/tree/master/timelpase)
Params used: AIThinker style, wcresolution 5, wcclock 20
Web: ~35fps

Multiple web sessions available. (with 2 running ~20 fps)
Multiple RTSP sessions available.
Video in main menu can be turned on/off, whilst still processing images.
Berry scripting support by default.

**Additional features (some new, some based on existing script related code)**
Motion available via TAS commands.
Image save to FS (flash or SD) available as TAS commands.
Berry triggers on frame(if enabled) and/or motion.
Image jpg decode and encode available as TAS commands (up to four slots for image memory - based on scripts implementation)
Images can be read from berry - jpeg or raw pixels (e.g. to post to website, manipulate, use with AI, etc.)
Camera can be disabled using `wcresolution -1`

**Advanced additional features:**
Motion detection tunable (scale of jpeg pixels configurable, jpeg size triggers, mask image may be used to mask out moving e.g. trees/road)

### thing that the new driver enables people to do:
timelapse with  mutliple possibilites
 - e.g. write to SD and view directly with static webserving enabled and a simple webpage on the SD card.
 - e.g. post images and data about them to a node-red flow, or other simple webserver.
 - berry could be used with ULP to deepsleep between pictures.  i.e. a battery powered timelapse becomes possible.
timelapse AND motion at the same time.  multiple timelapse configurations at the same time.
RTSP to other video, without the risk of the RTSP session terminating because of a second connection.
MJPEG to other video, without the risk of the MJPEG session terminating because you launch the web interface.
It enables the future use of AI image analysis.

**Reply by pkkrusty** [reply]
Thanks for all the work on this. I'll get a few modules flashed and play with them for a while.

**Comment by btsimonh** [comment]
It would be useful to understand all the existing use cases for ESP32cam - that could advise on the focus for any documentation changes and example Berry snippets, etc.

**Reply by tablatronix** [reply]
Are there any berry examples for this?

**Reply by sfromis** [reply]
With a quick search, I found a couple of places with a bit of code
https://tasmota.github.io/docs/Berry/#cam-module
https://tasmota.github.io/docs/Berry/#img-class
https://tasmota.github.io/docs/TFL/#berry-implementation
https://github.com/fknaopen/Berry_webcam_motion
https://github.com/tasmota/Berry_playground/blob/main/modules/googleapis/examples/gdrivepics.be

**Reply by btsimonh** [reply]
> Are there any berry examples for this?

also

https://github.com/btsimonh/TasmotaBerryPLay/tree/master/timelapse

check the issues page for a recent question ref current dev building of the updated driver & timelapse.

**Reply by tablatronix** [reply]
Thanks I have been trying to get getFrame and getStored working and there seems to be several  issues with the picstore buffer. So gonna just look into straight berry or a driver

**Comment by mrneutron42** [comment]
Hi, 
I'm curious if these fixes stop the InitErr 0x105 error after a soft reboot (restart 1) command?
reference: https://github.com/arendst/Tasmota/discussions/18492
```
09:09:53.637 CMD: wcinit
09:09:53.677 CAM: InitErr 0x105 try 1
09:09:53.714 CAM: InitErr 0x105 try 2
09:09:53.751 CAM: InitErr 0x105 try 3
09:09:53.752 CAM: InitErr 0x105
09:09:53.789 CAM: InitErr 0x105 try 1
09:09:53.826 CAM: InitErr 0x105 try 2
09:09:53.863 CAM: InitErr 0x105 try 3
09:09:53.864 CAM: InitErr 0x105
```

**Reply by sfromis** [reply]
IME, it does not make the bad hardware work.

**Reply by mrneutron42** [reply]
I did a custom compile and tested the V2 driver.  
I found that it does solve the InitErr 0x105 problems.

**Comment by tablatronix** [comment]
@mrneutron42  those look like camera init issues, check your gpio 0 or 16 on that board. My issue is that pic buffer does not allocate on non psram or my heap is too small not sure yet

**Comment by mrneutron42** [comment]
I've got some the the ESP32-CAM boards with the DM on the metal cover 
https://www.amazon.com/gp/product/B096WBMZ43/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&th=1
and one of them has been giving the InitErr 0x105 when I issue the 'wcinit' command.

I have the camera in a remote location and a 'restart 1' command is my only option of restarting it from a distance.
Doing the 'restart 1' soft restart has not been enough to recover the functionality.

This has happened in the past with this particular camera.  
I don't have any wires connected to the pin headers, except for 5V and GROUND.
I think my workaround was to physically pull the power plug to the camera and do a hard reboot.  
It recovered the camera for a few months, but the problem occurred again.

**Comment by tablatronix** [comment]
I just ordered some, is this the board that doesnt have a camera reset line? That would explain it, you would have to add some kind of pmic or watchdog to hard reset

**Reply by mrneutron42** [reply]
I think this is the schematic for the ESP32 camera board:
https://github.com/SeeedDocument/forum_doc/raw/master/reg/ESP32_CAM_V1.6.pdf
I see a FET connected to GPIO32 which switches the power to the CSI 2.8V and CSI 1.2V power rails.
So I think my boards would respond to the power switching GPIO.

**Comment by mrneutron42** [comment]
In this post https://github.com/arendst/Tasmota/discussions/18492#discussioncomment-6057352
btsimonh said he extended the reset period from 10ms to 200ms, which sounded like it helps.
@btsimonh, I'd like to try the revised firmware and see if it helps my cameras.

**Comment by mrneutron42** [comment]
I just made a trip to my remote camera site and simply did a 'hard reset' (pulled the power for 10 seconds) and the camera was recovered - no more 'InitErr 0x105' errors upon repower!  
To recap, doing a remote "soft reset" (issuing the 'restart 1' command) did not work.
Conclusion:  this is not a physical connection problem with my ESP32-CAM camera.

I suspect that the longer (200mS) power down in the modified code would help my recover my cameras, when they go stupid.

**Comment by mrneutron42** [comment]
I used TasmoCompiler with the #define USE_WEBCAM_V2 parameter and made a 14.4.1 tasmota32-webcam firmware with V2 driver.
It works fine with the "DM" ESP32-CAM modules I have on my desk.
I just loaded it into one of my remote cameras and it is working fine.

Is this V2 driver eventually going to replace the V1 driver, and thus we won't have to make custom compiles to get the fixes and improvements of the V2 driver?

**Comment by tablatronix** [comment]
Yeah I think a nightly binary for this would be nice, I am now using this on all my cameras

**Reply by btsimonh** [reply]
@sfromis @Jason2866 - how about it?

**Reply by sfromis** [reply]
Maybe the question is if the regular precompiled webcam build should be V2 instead, but with my low usage I can only say that my horrible experience with seeing webcam occasionally work has not gotten even worse with V2.

**Reply by btsimonh** [reply]
I did spend a lot of time on it, and in general terms the performance is better (higher FPS, more stable, etc.).  I was hoping for more feedback, and that based on that we could mainstream it.  But, I never tried it with audio (bought the hardware, maybe I'll return to audio with video in the summer).  I think if initially we ADDED a config (or two, one for normal esp32cam, and one for S3), then we'd get more direct feedback because people could have the choice.

**Reply by Jason2866** [reply]
@arendst Had no issues with the v2 Webcam build. We could make default and let us see how it goes. What's your opinion?

**Reply by arendst** [reply]
I'll do some tests today and if succesful will enable v2 asap

**Reply by arendst** [reply]
OK. Works on both AIThinker and TTGO Camera. Just noticed too much logging:
```
09:58:16.646 CAM: avFPS 40.016006 jpg FS:12924(0) f:1382 s:48
09:58:19.145 CAM: avFPS 40.016006 jpg FS:13607(0) f:1426 s:48
09:58:21.644 CAM: avFPS 40.016006 jpg FS:17183(0) f:1468 s:48
09:58:24.143 CAM: avFPS 40.016006 jpg FS:17721(0) f:1505 s:48
09:58:26.642 CAM: avFPS 40.016006 jpg FS:18465(0) f:1547 s:48
09:58:29.139 CAM: avFPS 40.016006 jpg FS:12082(0) f:1585 s:48
09:58:31.639 CAM: avFPS 40.016006 jpg FS:19121(0) f:1626 s:48
09:58:34.138 CAM: avFPS 40.016006 jpg FS:18525(0) f:1661 s:48
09:58:36.636 CAM: avFPS 40.016006 jpg FS:18316(0) f:1708 s:48
09:58:39.135 CAM: avFPS 40.016006 jpg FS:18339(0) f:1752 s:48
09:58:41.634 CAM: avFPS 40.016006 jpg FS:18976(0) f:1790 s:48
09:58:44.134 CAM: avFPS 40.016006 jpg FS:18690(0) f:1823 s:48
```
I'll change this now before release.

**Reply by mrneutron42** [reply]
That's great to hear!  Thanks @arendst!

**Reply by mrneutron42** [reply]
@arendst if you have time before the next release, it would be good to correct an error in the comments sections of the in the xdrv...webcam.ino files.  
WcAECLevel is an invalid command name.  The actual command name in the code is WcAELevel
```
 * WcAECLevel	  = Auto exposure control level, -2 ... +2
 change to
 * WcAELevel      = Auto exposure control level, -2 ... +2
```
I see it in:
tasmota/tasmota_xdrv_driver/xdrv_81_esp32_webcam_task.ino
tasmota/tasmota_xdrv_driver/xdrv_81_esp32_webcam.ino

**Reply by sfromis** [reply]
Strictly speaking, the error arguably is that the command name *should* have been WcAECLevel, but fixing this typo would be a bit of a breaking change.

**Reply by mrneutron42** [reply]
It sounds like you are saying that the command name down in the code has been spelled wrong, for a long time?  (thus it doesn't match the command name in the comment sections of the .ino files)

**Reply by sfromis** [reply]
What I mean is that the general abbreviation, like used in the related command names for automatic exposure, is "AEC" for automatic exposure control.

OTOH, as the data structure internally does use the field name `ae_level` making sense for command name without the "C", I'm not going to be hard-nosed about this inconsistency :wink:

**Reply by mrneutron42** [reply]
Yes, it would have been nice to have the variable name line up with the actual function (AEC).
If its too painful to correct the variable name mistake, it would be nice for the documentation to let users know what the actual command they need to send is.  I stumbled on the correct command by taking shots in the dark until....hey, that worked.
How about this?
```
 * WcAECLevel	  = (intended command name) Auto exposure control level, -2 ... +2)
 * WcAELevel     = (actual command name) Auto exposure control level, -2 ... +2)
```

**Comment by Jason2866** [comment]
Crappy esp32cam modules will still not work, software can't repair hardware design flaws.

**Reply by btsimonh** [reply]
yeah, but some of them are not-so-crappy, but just missing wires (like the DM ones).  But there ARE some compromised boards out there where the hardware design does not take account of the bandwidths required.  e.g. where the CAM frequency and Wifi interact badly, due to bad board layout/design (well, I have at least one where changing the cam freq significantly improved wifi.. which should not happen).

**Comment by tablatronix** [comment]
I did have some issues with nopsram modules, I think I patched it like this, but I wound up replacing these with esp32-cams and abandoning that dev board.

![image](https://github.com/user-attachments/assets/9bf3c316-24f5-48bf-a152-d3ef7260f1c1)

**Comment by mrneutron42** [comment]
As a side question, is there a place that still sells the ESP32-CAM boards made by AI Thinker?
The ESP32-CAMs on Amazon.com all show pictures with the ESP logo on the modules, but when you get them they have the DM logo on them.

**Reply by btsimonh** [reply]
I bought an S3 based board that is pretty nice (and more powerful).  I also got a XIAO ESP32S3 Sense with camera.  I've not run it in anger yet, and it's so small that I would worry about heat (they have a heat disclaimer on the website).  But it sure looks nice and has type-C for power - and also has audio?

**Reply by sfromis** [reply]
While the Seeed Studio XIAO ESP32S3 Sense supports I2S, it has no speaker port, unless you count PWM to drive a tiny speaker/beeper.

Using I2S, you can still connect a amplifier chip like MAX98357 facilitating things like a MP3 player in decent quality, etc. MEMS microphones are also an option. Cheap breakout boards are available. Some boards do integrate this, but not those tiny ESP32-S3 boards.

**Reply by btsimonh** [reply]
in-built 'digital mic', but no output device.

**Reply by Jason2866** [reply]
Do not recommend the Seeed Studio XIAO ESP32S3 sense. Probably design related wifi issues.

**Reply by tablatronix** [reply]
Thats a shame, they aren't cheap either, but love how tiny

**Reply by sfromis** [reply]
I've seen antenna trouble with other modules in that tiny form factor. An antenna just works better in a less cramped space.

**Comment by mrneutron42** [comment]
@btsimonh @arendst  I tried development build 14.5.0.3 from https://ota.tasmota.com/tasmota32/
which is supposed to have the new V2 webcam driver active in it, as of 14.5.0.2.
* ESP32 enable webcam version 2 (#18732)

Unfortunately, it is not behaving like (doesn't fix the init errors after a soft reset) like the custom compiles of 14.5.0 I've made with TasmoCompiler using the #define USE_WEBCAM_V2 parameter.
I am able to eliminate the init errors by again loading my custom compiled Tasmota32 14.5.0 firmware.
I'm guessing that there is some difference between your test version of xdrv_81_esp32_webcam_task.ino and the version in the Tasmota development branch?

@btsimonh you said:
```
I think my final fix was toggling the power on the cam module, around line 1020 of
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_81_esp32_webcam_task.ino
```
In the section with "  // cannot hurt to retry..." I see a difference, but I'm not sure if this is the problem.

https://github.com/btsimonh/Tasmota/blob/webcam2023/tasmota/tasmota_xdrv_driver/xdrv_81_esp32_webcam_task.ino
```
      if (err == 0x105){
        // try a longer power off... and retry
```
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_81_esp32_webcam_task.ino
```
      if (err == 0x105 && (config.pin_pwdn >= 0)){
        // try a longer power off... and retry
```
